### DISCUSSION TITLE
Searching for the code module

### DISCUSSION THREAD

**Discussion by iv3wjr** [micropython/micropython#13550] [Libraries & Drivers]
Hi to all,

  for many years my favorite Python debugging tool has been the embed() function from the IPython package.  Only recently (my fault!) I discovered the existence of the [code](https://docs.python.org/3/library/code.html) module.
  I've always been looking for a similar module for Micropython but I could not find it. Is there anyone who can point me to someone working on it or explain why this is not yet available in the standard library?

**Comment by andrewleech** [comment]
The micropython repl isn't broken into functions that match the cpython code module, nor are the internals exposed to python in a similar way. As with many similar questions about micropython vs cpython functionality, the reason for this is presumably both to reduce code size and because no one has ever needed it exposed to python enough to implement it.

You may be interested in aiorepl in micropython-lib though? It's essentially a repl written in pure python.

**Reply by iv3wjr** [reply]
Andrew, thank you for your kind answer. I had a look at aiorepl but its usage is not so straightforward as
```
from IPython import embed; embed()
```
which is my favorite way of inspecting what is wrong with my code.
Playing with the [usage example](https://github.com/micropython/micropython-lib/tree/master/micropython/aioreplurl) I learned something  new for me: in Micropython local variables are *not* included in the dictionary returned by the locals() function, so there is no way of accessing them from a REPL. That said, I think that I must search an alternative way to root around in my buggy code.

**Comment by Josverl** [comment]
If you like IPython/Jupyter, then perhaps [micropython-magic](https://github.com/Josverl/micropython-magic) may be of use to you.
I created it to experiment with micropython, without the need to setup a project for each port/ board, or worry about uploading files too much

**Reply by andrewleech** [reply]
Oh interesting, I made one of these some time back too :-D it'd be interesting to compare notes! https://gitlab.com/alelec/jupyter_micropython_remote

**Reply by Josverl** [reply]
Yes, I found that one, and created a kernel initially myself. Then found out that to get visualizations and all the other goodness required a huge effort.
That's when I found the custom magics that allow that with zero effort. As in all choices there was effort to pay in other areas...
I also look forward to use other protocols when available in mpremote
